### ISSUE TITLE
Secure Tunelling is not working

### ISSUE THREAD

**Issue by afreyermuth98** [awslabs/aws-iot-device-client#223]
**Describe the bug**

When I launch a basic aws iot device client with every feature at false, there are no problem but when I set secure tunelling to true, the connection resets non stop. 

But then, I'm trying to activate secure tunneling to create an AWS IoT  tunnel and it seems that the device is connecting and deconnecting non stop to my MQTT broker. Here is the error message I'm getting : 

**To Reproduce**

1. Reproduce all the basic steps. In setup.sh script, say no to everything except Secure Tunneling
2. Launch the client : ./aws-iot-device-client

**Expected behavior**
In logs : 
`Subscribed to tunnel notification topic`

**Actual behavior**

A clear and concise description of what actually happened.

**Logs**

The last 2 lines of logs are repeating non stop.
```
2022-03-02T17:26:45.922Z [INFO]  {SharedCrtResourceManager.cpp}: Starting Device Client features.
2022-03-02T17:26:45.922Z [INFO]  {SecureTunnelingFeature.cpp}: Running Secure Tunneling!
2022-03-02T17:26:45.922Z [INFO]  {Main.cpp}: Client base has been notified that Secure Tunneling has started
2022-03-02T17:26:45.925Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `libaws-c-mqtt: AWS_ERROR_MQTT_UNEXPECTED_HANGUP, The connection was closed unexpectedly.`. Device Client will retry connection until it is successfully connected to the core. 
2022-03-02T17:26:46.086Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection resumed with return code: 0
2022-03-02T17:26:46.089Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `libaws-c-mqtt: AWS_ERROR_MQTT_UNEXPECTED_HANGUP, The connection was closed unexpectedly.`. Device Client will retry connection until it is successfully connected to the core. 
2022-03-02T17:26:46.237Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection resumed with return code: 0
2022-03-02T17:26:46.240Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `libaws-c-mqtt: AWS_ERROR_MQTT_UNEXPECTED_HANGUP, The connection was closed unexpectedly.`. Device Client will retry connection until it is successfully connected to the core. 
```

**Environment (please complete the following information):**
 - OS: Ubuntu
 - Version 20
 - Architecture: x86-64

**Comment by shangabl** [comment]
Hi @afreyermuth98,

Thank you for reaching out.

Are you running Device Client as a service? It is possible that Device Client is already running which is causing the connection errors. Before starting try running `systemctl status aws-iot-device-client.service` to see if it is already active. If so, you can stop it with `systemctl stop aws-iot-device-client.service` and try again.

Let me know if that works! Happy to help if it does not.

Shane Gable

**Comment by afreyermuth98** [comment]
Hello @shangabl 

When I've configured my IoT device client, I've answered "no" to Device Client as a service. 
I've still tried to run `systemctl status aws-iot-device-client.service` and as I could expected I got : 
`Unit aws-iot-device-client.service could not be found.`

Anthony

**Comment by shangabl** [comment]
Hi Anthony,

Can you also verify that your certificate is active and attached to the Thing? You can check this in the AWS Console by going to IoT Core -> Manage -> Things -> (Click your Thing) -> Certificates Tab

Please confirm there is a certificate there and that is active. If so, can you also confirm by clicking on the certificate that it has a Policy attached.

Shane

**Comment by afreyermuth98** [comment]
Hi Shane,

Yes it is, when the option of secure tunneling is not activate, it works fine, when I try with secure tunelling, it fails as I described :/
Moreover it has a policy attached with "Allow" at iot:Connect, iot:Publish, iot:Subscribe and iot:Receive

Anthony

iot:Publish
--


<br class="Apple-interchange-newline">

iot:Publish
--


<br class="Apple-interchange-newline">

iot:Publish
--


<br class="Apple-interchange-newline">

iot:Publish
--


<br class="Apple-interchange-newline">

**Comment by shangabl** [comment]
Hi Anthony,

Sorry I overlooked the fact that the client may be running but not as a service... Can you check that it isn't already running as a process?

`ps -ef | grep aws-iot-device-client`

Thanks
Shane

**Comment by afreyermuth98** [comment]
Hello Shane,

It's running as a process indeed. 

`user@instance:~$ ps -ef | grep aws-iot-device-client
user     37124       1  0 Mar02 ?        00:00:05 ./aws-iot-device-client`

I've killed it and relaunched the client with secure tunelling activated and got the same results :'(

Sincerely

Anthony

**Comment by shangabl** [comment]
Hi Anthony,

I tried to recreate the issue but did not see the same result. Can you give me all the logs you can and your configuration file so I can do my best to recreate the issue? Please remove your endpoint from the configuration file before posting it though and also be careful not to post any private information. That should really help me figure out the issue.

Thanks!

Shane

**Comment by afreyermuth98** [comment]
Hello Shane,

Here is the logs I get when I start the device client. You can also find the config file inside the logs when it displays it. As you'll see, I've installed the device as "ubuntu" user, if it can helps.

Thanks a lot.

Anthony

```
2022-03-07T08:52:16.289Z [WARN]  {Config.cpp}: Key {template-name} was provided in the JSON configuration file with an empty value
2022-03-07T08:52:16.289Z [WARN]  {Config.cpp}: Key {csr-file} was provided in the JSON configuration file with an empty value
2022-03-07T08:52:16.289Z [WARN]  {Config.cpp}: Key {device-key} was provided in the JSON configuration file with an empty value
2022-03-07T08:52:16.289Z [WARN]  {Config.cpp}: Key {publish-topic} was provided in the JSON configuration file with an empty value
2022-03-07T08:52:16.289Z [WARN]  {Config.cpp}: Key {subscribe-topic} was provided in the JSON configuration file with an empty value
2022-03-07T08:52:16.289Z [WARN]  {Config.cpp}: Shadow Name {shadow-name} was provided in the JSON configuration file with an empty value
2022-03-07T08:52:16.289Z [WARN]  {Config.cpp}: Input file {shadow-input-file} was provided in the JSON configuration file with an empty value
2022-03-07T08:52:16.289Z [WARN]  {Config.cpp}: Output file {shadow-output-file} was provided in the JSON configuration file with an empty value
2022-03-07T08:52:16.289Z [INFO]  {Config.cpp}: Successfully fetched JSON config file: 

    {
      "endpoint":	"<MY_ENDPOINT>.iot.eu-central-1.amazonaws.com",
      "cert":	"/home/ubuntu/certs/cert.pem",
      "key":	"/home/ubuntu/certs/private.pem.key",
      "root-ca":	"/home/ubuntu/certs/rootca.pem",
      "thing-name":	"<DEVICE_NAME>",
      "logging":	{
        "level":	"DEBUG",
        "type":	"STDOUT",
        "file": "/var/log/aws-iot-device-client/aws-iot-device-client.log",
        "enable-sdk-logging":	false,
        "sdk-log-level":	"TRACE",
        "sdk-log-file": "/var/log/aws-iot-device-client/sdk.log"
      },
      "jobs":	{
        "enabled":	false,
        "handler-directory": "/home/ubuntu/.aws-iot-device-client/jobs"
      },
      "tunneling":	{
        "enabled":	true
      },
      "device-defender":	{
        "enabled":	false,
        "interval": 300
      },
      "fleet-provisioning":	{
        "enabled":	false,
        "template-name": "",
        "template-parameters": "",
        "csr-file": "",
        "device-key": ""
      },
      "samples": {
        "pub-sub": {
          "enabled": false,
          "publish-topic": "",
          "publish-file": "/home/ubuntu/.aws-iot-device-client/pubsub/publish-file.txt",
          "subscribe-topic": "",
          "subscribe-file": "/home/ubuntu/.aws-iot-device-client/pubsub/subscribe-file.txt"
        }
      },
      "config-shadow":	{
        "enabled":	false
      },
      "sample-shadow": {
        "enabled": false,
        "shadow-name": "",
        "shadow-input-file": "",
        "shadow-output-file": ""
      }
    }

2022-03-07T08:52:16.289Z [DEBUG] {Config.cpp}: Did not find a runtime configuration file, assuming Fleet Provisioning has not run for this device
2022-03-07T08:52:16.289Z [DEBUG] {EnvUtils.cpp}: Updated PATH environment variable to: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/ubuntu/.aws-iot-device-client:/home/ubuntu/.aws-iot-device-client/jobs:/home/ubuntu/aws-iot-device-client/build:/home/ubuntu/aws-iot-device-client/build/jobs
2022-03-07T08:52:16.289Z [INFO]  {Main.cpp}: Now running AWS IoT Device Client version v1.5.9-5cc4d24
2022-03-07T08:52:16.293Z [INFO]  {SharedCrtResourceManager.cpp}: SDK logging is disabled. Enable it with --enable-sdk-logging on the command line or logging::enable-sdk-logging in your configuration file
2022-03-07T08:52:16.293Z [DEBUG] {Retry.cpp}: Retryable function starting, it will retry until success
2022-03-07T08:52:16.298Z [INFO]  {SharedCrtResourceManager.cpp}: Establishing MQTT connection with client id AWS_000262...
2022-03-07T08:52:16.445Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection established with return code: 0
2022-03-07T08:52:16.446Z [INFO]  {SharedCrtResourceManager.cpp}: Shared MQTT connection is ready!
2022-03-07T08:52:16.446Z [INFO]  {Main.cpp}: Jobs is disabled
2022-03-07T08:52:16.446Z [INFO]  {Main.cpp}: Secure Tunneling is enabled
2022-03-07T08:52:16.446Z [INFO]  {Main.cpp}: Device Defender is disabled
2022-03-07T08:52:16.446Z [INFO]  {Main.cpp}: Sample shadow is disabled
2022-03-07T08:52:16.446Z [INFO]  {Main.cpp}: Pub Sub is disabled
2022-03-07T08:52:16.446Z [INFO]  {SharedCrtResourceManager.cpp}: Starting Device Client features.
2022-03-07T08:52:16.446Z [INFO]  {SecureTunnelingFeature.cpp}: Running Secure Tunneling!
2022-03-07T08:52:16.446Z [INFO]  {Main.cpp}: Client base has been notified that Secure Tunneling has started
2022-03-07T08:52:16.448Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `libaws-c-mqtt: AWS_ERROR_MQTT_UNEXPECTED_HANGUP, The connection was closed unexpectedly.`. Device Client will retry connection until it is successfully connected to the core. 
2022-03-07T08:52:16.597Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection resumed with return code: 0
2022-03-07T08:52:16.600Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `libaws-c-mqtt: AWS_ERROR_MQTT_UNEXPECTED_HANGUP, The connection was closed unexpectedly.`. Device Client will retry connection until it is successfully connected to the core. 
2022-03-07T08:52:16.756Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection resumed with return code: 0
2022-03-07T08:52:16.759Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `libaws-c-mqtt: AWS_ERROR_MQTT_UNEXPECTED_HANGUP, The connection was closed unexpectedly.`. Device Client will retry connection until it is successfully connected to the core. 
2022-03-07T08:52:16.904Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection resumed with return code: 0
2022-03-07T08:52:16.907Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `libaws-c-mqtt: AWS_ERROR_MQTT_UNEXPECTED_HANGUP, The connection was closed unexpectedly.`. Device Client will retry connection until it is successfully connected to the core. 
2022-03-07T08:52:17.064Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection resumed with return code: 0
2022-03-07T08:52:17.068Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `libaws-c-mqtt: AWS_ERROR_MQTT_UNEXPECTED_HANGUP, The connection was closed unexpectedly.`. Device Client will retry connection until it is successfully connected to the core. 
2022-03-07T08:52:17.221Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection resumed with return code: 0
2022-03-07T08:52:17.224Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `libaws-c-mqtt: AWS_ERROR_MQTT_UNEXPECTED_HANGUP, The connection was closed unexpectedly.`. Device Client will retry connection until it is successfully connected to the core. 
2022-03-07T08:52:17.376Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection resumed with return code: 0
2022-03-07T08:52:17.380Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `libaws-c-mqtt: AWS_ERROR_MQTT_UNEXPECTED_HANGUP, The connection was closed unexpectedly.`. Device Client will retry connection until it is successfully connected to the core. 


```

**Comment by shangabl** [comment]
Thanks Anthony! I'll do my best here to recreate the issue. As reply as soon as I have any new info.

**Comment by shangabl** [comment]
Hi Anthony,

I attempted to recreate the issue using the same configuration that you provided and it worked as expected. Are you using a your own MQTT broker?

**Comment by afreyermuth98** [comment]
Hello Shane,

No, I'm using an AWS MQTT Broker. All my configuration is on AWS (MQTT & EC2).
Maybe I can make you a step by step slide show of my configuration and show it to you ?

Sincerely,

Anthony

**Comment by satheshantanu** [comment]
Hi @afreyermuth98 - can you please provide your email ID? We'd be happy to contact you over a call.

**Comment by afreyermuth98** [comment]
Hello !

Sure, you can contact me at anthofreyer@gmail.com

Sincerely.

**Comment by shangabl** [comment]
Hi Anthony,

Thank you for taking the time to meet with us! We really appreciate it.

So, as you know we figured out that this was a permissions issue and that the resources were scoped too narrowly. We have a least privilege policy for a device using the Device Client with the Secure Tunneling feature enabled below:
```
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "iot:Connect",
      "Resource": "arn:aws:iot:<region>:<accountId>:client/<thingName>"
    },
    {
      "Effect": "Allow",
      "Action": "iot:Receive",
      "Resource": "arn:aws:iot:<region>:<accountId>:topic/<thingName>"
    },
    {
      "Effect": "Allow",
      "Action": "iot:Publish",
      "Resource": "arn:aws:iot:<region>:<accountId>:topic/<thingName>"
    },
    {
      "Effect": "Allow",
      "Action": "iot:Subscribe",
      "Resource": "arn:aws:iot:<region>:<accountId>:topicfilter/<thingName>"
    },
    {
      "Effect": "Allow",
      "Action": "iot:Subscribe",
      "Resource": "arn:aws:iot:<region>:<accountId>:topicfilter/$aws/things/<thingName>/tunnels/notify"
    },
    {
      "Effect": "Allow",
      "Action": "iot:Receive",
      "Resource": "arn:aws:iot:<region>:<accountId>:topic/$aws/things/<thingName>/tunnels/notify"
    }
  ]
}
```

**Comment by arunreddy00** [comment]
Hi Shane,

I'm facing same issue with secure tunneling. There is 5minutes interval gap on every retry to establish the connection with tunnel.

Can you help me in solving the issue.

Please find below logs 
```
Mar 24 17:31:06 desktop bash[31171]: 2022-03-24T17:31:06.606Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `aws-c-io: AWS_IO_SOCKET_CLOSED, socket is closed.`. Device Client will retry connection until it is successfully connected to the core.
Mar 24 17:31:07 desktop bash[31171]: 2022-03-24T17:31:07.636Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection resumed with return code: 0
Mar 24 17:36:08 desktop bash[31171]: 2022-03-24T17:36:08.357Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `aws-c-io: AWS_IO_SOCKET_CLOSED, socket is closed.`. Device Client will retry connection until it is successfully connected to the core.
Mar 24 17:36:09 desktop bash[31171]: 2022-03-24T17:36:09.343Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection resumed with return code: 0
Mar 24 17:41:10 desktop bash[31171]: 2022-03-24T17:41:10.304Z [ERROR] {SharedCrtResourceManager.cpp}: MQTT Connection interrupted with error: `aws-c-io: AWS_IO_SOCKET_CLOSED, socket is closed.`. Device Client will retry connection until it is successfully connected to the core.
```

**Comment by shangabl** [comment]
Hi arunreddy00,

I would love to help!

Can you first check that you are not running duplicate clients? That is one way that this can happen. `ps -ef | grep aws-iot-device-client` should reveal if you already have a process running. If so, please kill the process before trying again. 

I would then ask you to confirm you have the correct certificates and keys configured. Please confirm that your device certificate is active with a policy attached. The file path to this certificate should be given as the value of "cert" in the configuration file. Also, make sure that the private key and AmazonRootCA are correct.

If all that is correct, please check that the policy attached to the device certificate has at least these permissions
```
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "iot:Connect",
      "Resource": "arn:aws:iot:aws-region:AWS-account-ID:client/${iot:Connection.Thing.ThingName}"
    },
    {
      "Effect": "Allow",
      "Action": "iot:Subscribe",
      "Resource": "arn:aws:iot:aws-region:AWS-account-ID:topicfilter/$aws/things/${iot:Connection.Thing.ThingName}/tunnels/notify"
    },
    {
      "Effect": "Allow",
      "Action": "iot:Receive",
      "Resource": "arn:aws:iot:aws-region:AWS-account-ID:topic/$aws/things/${iot:Connection.Thing.ThingName}/tunnels/notify"
    }
  ]
}
```

If after trying that the issue is not resolved then please share your configuration file here. However, make sure to remove your personal endpoint so it is not shared publicly. Then I'll take a look and see what I can do.

Thanks!
Shane

**Comment by arunreddy00** [comment]
Hi Shane

Thank you for the quick response.

Things, certs folder and certificate with permissions all are good.

I also established secure tunnel from IOT to edge device in 5min interval. But unfortunately it disconnects due to 5minutes interval gap on every retry to establish the connection with tunnel MQTT.


Please find below config file with logs from tunnel service.
```
Mar 24 18:09:06 desktop bash[25730]: 2022-03-24T18:09:06.787Z [INFO]  {Config.cpp}: Successfully fetched JSON 
Mar 24 18:09:06 desktop bash[25730]:     {
Mar 24 18:09:06 desktop bash[25730]:       "endpoint":       "IOT_endpoint"
Mar 24 18:09:06 desktop bash[25730]:       "cert":   "/etc/certs/mqtt.certificate.pem",
Mar 24 18:09:06 desktop bash[25730]:       "key":    "/etc/certs/mqtt.private.key",
Mar 24 18:09:06 desktop bash[25730]:       "root-ca":        "/etc/ssl/certs/Amazon_Root_CA_1.pem",
Mar 24 18:09:06 desktop bash[25730]:       "thing-name":     "thing_name",
Mar 24 18:09:06 desktop bash[25730]:       "logging":        {
Mar 24 18:09:06 desktop bash[25730]:         "level":        "DEBUG",
Mar 24 18:09:06 desktop bash[25730]:         "type": "STDOUT",
Mar 24 18:09:06 desktop bash[25730]:         "file": "/var/log/aws-iot-device-client/aws-iot-device-client.log"
Mar 24 18:09:06 desktop bash[25730]:         "enable-sdk-logging":   false,
Mar 24 18:09:06 desktop bash[25730]:         "sdk-log-level":        "TRACE",
Mar 24 18:09:06 desktop bash[25730]:         "sdk-log-file": "/var/log/aws-iot-device-client/sdk.log"
Mar 24 18:09:06 desktop bash[25730]:       },
Mar 24 18:09:06 desktop bash[25730]:       "jobs":   {
Mar 24 18:09:06 desktop bash[25730]:         "enabled":      false,
Mar 24 18:09:06 desktop bash[25730]:         "handler-directory": "/home/user/.aws-iot-device-client/jobs"
Mar 24 18:09:06 desktop bash[25730]:       },
Mar 24 18:09:06 desktop bash[25730]:       "tunneling":      {
Mar 24 18:09:06 desktop bash[25730]:         "enabled":      true
Mar 24 18:09:06 desktop bash[25730]:       },
Mar 24 18:09:06 desktop bash[25730]:       "device-defender":        {
Mar 24 18:09:06 desktop bash[25730]:         "enabled":      false,
Mar 24 18:09:06 desktop bash[25730]:         "interval": 3000
Mar 24 18:09:06 desktop bash[25730]:       },
Mar 24 18:09:06 desktop bash[25730]:       "fleet-provisioning":     {
Mar 24 18:09:06 desktop bash[25730]:         "enabled":      false,
Mar 24 18:09:06 desktop bash[25730]:         "template-name": "",
Mar 24 18:09:06 desktop bash[25730]:         "template-parameters": "",
Mar 24 18:09:06 desktop bash[25730]:         "csr-file": "",
Mar 24 18:09:06 desktop bash[25730]:         "device-key": ""
Mar 24 18:09:06 desktop bash[25730]:       },
Mar 24 18:09:06 desktop bash[25730]:       "samples": {
Mar 24 18:09:06 desktop bash[25730]:         "pub-sub": {
Mar 24 18:09:06 desktop bash[25730]:           "enabled": false,
Mar 24 18:09:06 desktop bash[25730]:           "publish-topic": "",
Mar 24 18:09:06 desktop bash[25730]:           "publish-file": "/home/user/.aws-iot-device-client/pubsub/pu
Mar 24 18:09:06 desktop bash[25730]:           "subscribe-topic": "",
Mar 24 18:09:0 desktop bash[25730]:           "subscribe-file": "/home/user/.aws-iot-device-client/pubsub/
Mar 24 18:09:06 desktop bash[25730]:         }
Mar 24 18:09:06 desktop bash[25730]:       },
Mar 24 18:09:06 desktop bash[25730]:       "config-shadow":  {
Mar 24 18:09:06 desktop bash[25730]:         "enabled":      false
Mar 24 18:09:06 desktop bash[25730]:       },
Mar 24 18:09:06 desktop bash[25730]:       "sample-shadow": {
Mar 24 18:09:06 desktop bash[25730]:         "enabled": false,
Mar 24 18:09:06 desktop bash[25730]:         "shadow-name": "",
Mar 24 18:09:06  desktop bash[25730]:         "shadow-input-file": "",
Mar 24 18:09:06 desktop bash[25730]:         "shadow-output-file": ""
Mar 24 18:09:06 desktop bash[25730]:       }
Mar 24 18:09:06 desktop bash[25730]:     }
Mar 24 18:09:06 desktop bash[25730]: 2022-03-24T18:09:06.788Z [DEBUG] {Config.cpp}: Did not find a runtime conf
Mar 24 18:09:06 desktop bash[25730]: 2022-03-24T18:09:06.789Z [DEBUG] {EnvUtils.cpp}: Updated PATH environment
Mar 24 18:09:06 desktop bash[25730]: 2022-03-24T18:09:06.789Z [INFO]  {Main.cpp}: Now running AWS IoT Device Cl
Mar 24 18:09:06 desktop bash[25730]: 2022-03-24T18:09:06.795Z [INFO]  {SharedCrtResourceManager.cpp}: SDK loggi
Mar 24 18:09:06 desktop bash[25730]: 2022-03-24T18:09:06.796Z [DEBUG] {Retry.cpp}: Retryable function starting,
Mar 24 18:09:06 desktop bash[25730]: 2022-03-24T18:09:06.815Z [INFO]  {SharedCrtResourceManager.cpp}: Establish
Mar 24 18:09:07 desktop bash[25730]: 2022-03-24T18:09:07.762Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT conn
Mar 24 18:09:07 desktop bash[25730]: 2022-03-24T18:09:07.762Z [INFO]  {SharedCrtResourceManager.cpp}: Shared MQ
Mar 24 18:09:07 desktop bash[25730]: 2022-03-24T18:09:07.763Z [INFO]  {Main.cpp}: Jobs is disabled
Mar 24 18:09:07 desktop bash[25730]: 2022-03-24T18:09:07.763Z [INFO]  {Main.cpp}: Secure Tunneling is enabled
Mar 24 18:09:07 desktop bash[25730]: 2022-03-24T18:09:07.763Z [INFO]  {Main.cpp}: Device Defender is disabled
Mar 24 18:09:07 desktop bash[25730]: 2022-03-24T18:09:07.763Z [INFO]  {Main.cpp}: Sample shadow is disabled
Mar 24 18:09:07 desktop bash[25730]: 2022-03-24T18:09:07.763Z [INFO]  {Main.cpp}: Pub Sub is disabled
Mar 24 18:09:07 desktop bash[25730]: 2022-03-24T18:09:07.763Z [INFO]  {SharedCrtResourceManager.cpp}: Starting
Mar 24 18:09:07 desktop bash[25730]: 2022-03-24T18:09:07.763Z [INFO]  {SecureTunnelingFeature.cpp}: Running Sec
Mar 24 18:09:07 desktop bash[25730]: 2022-03-24T18:09:07.763Z [INFO]  {Main.cpp}: Client base has been notified
Mar 24 18:09:08 desktop bash[25730]: 2022-03-24T18:09:08.013Z [DEBUG] {SecureTunnelingFeature.cpp}: Subscribed
```

Thanks

**Comment by shangabl** [comment]
I think this is a separate issue from the original one. Would you mind opening a separate GitHub issue and so we can start tracking that separately?

**Comment by afreyermuth98** [comment]
Hello Shane !

First, I would like to thank you again for the time you got for me and for your help. I tried to add the policy you send to me and then launched my client with secure tunneling enabled. It seems to work (no error), but then when I create the tunnel inside AWS, I got no evolution on my iot client and it's "DISCONNECTED" on my tunnel in AWS console. Have you an idea of what it can be ?
Here are my logs when I launch my client and creating tunnel under AWS console :
`2022-03-25T13:32:31.107Z [WARN]  {Config.cpp}: Key {template-name} was provided in the JSON configuration file with an empty value
2022-03-25T13:32:31.107Z [WARN]  {Config.cpp}: Key {csr-file} was provided in the JSON configuration file with an empty value
2022-03-25T13:32:31.107Z [WARN]  {Config.cpp}: Key {device-key} was provided in the JSON configuration file with an empty value
2022-03-25T13:32:31.107Z [WARN]  {Config.cpp}: Key {publish-topic} was provided in the JSON configuration file with an empty value
2022-03-25T13:32:31.107Z [WARN]  {Config.cpp}: Key {subscribe-topic} was provided in the JSON configuration file with an empty value
2022-03-25T13:32:31.107Z [WARN]  {Config.cpp}: Shadow Name {shadow-name} was provided in the JSON configuration file with an empty value
2022-03-25T13:32:31.107Z [WARN]  {Config.cpp}: Input file {shadow-input-file} was provided in the JSON configuration file with an empty value
2022-03-25T13:32:31.107Z [WARN]  {Config.cpp}: Output file {shadow-output-file} was provided in the JSON configuration file with an empty value
2022-03-25T13:32:31.107Z [INFO]  {Config.cpp}: Successfully fetched JSON config file:     {
      "endpoint":	"<IOT_ENDPOINT>",
      "cert":	"/home/ubuntu/certs/certificate.pem.crt",
      "key":	"/home/ubuntu/certs/private.pem.key",
      "root-ca":	"/home/ubuntu/certs/root-ca.pem",
      "thing-name":	"<DEVICE_NAME>",
      "logging":	{
        "level":	"DEBUG",
        "type":	"STDOUT",
        "file": "/var/log/aws-iot-device-client/aws-iot-device-client.log"
      },
      "jobs":	{
        "enabled":	false,
        "handler-directory": "/root/.aws-iot-device-client/jobs"
      },
      "tunneling":	{
        "enabled":	true
      },
      "device-defender":	{
        "enabled":	false,
        "interval": 300
      },
      "fleet-provisioning":	{
        "enabled":	false,
        "template-name": "",
        "template-parameters": "",
        "csr-file": "",
        "device-key": ""
      },
      "samples": {
        "pub-sub": {
          "enabled": false,
          "publish-topic": "",
          "publish-file": "/root/.aws-iot-device-client/pubsub/publish-file.txt",
          "subscribe-topic": "",
          "subscribe-file": "/root/.aws-iot-device-client/pubsub/subscribe-file.txt"
        }
      },
      "config-shadow":	{
        "enabled":	false
      },
      "sample-shadow": {
        "enabled": false,
        "shadow-name": "",
        "shadow-input-file": "",
        "shadow-output-file": ""
      }
    }


2022-03-25T13:32:31.107Z [DEBUG] {Config.cpp}: Did not find a runtime configuration file, assuming Fleet Provisioning has not run for this device
2022-03-25T13:32:31.107Z [DEBUG] {EnvUtils.cpp}: Updated PATH environment variable to: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/ubuntu/.aws-iot-device-client:/home/ubuntu/.aws-iot-device-client/jobs:/home/ubuntu/aws-iot-device-client/build:/home/ubuntu/aws-iot-device-client/build/jobs
2022-03-25T13:32:31.107Z [INFO]  {Main.cpp}: Now running AWS IoT Device Client version v1.5.13-b6080e5
2022-03-25T13:32:31.108Z [INFO]  {SharedCrtResourceManager.cpp}: SDK logging is disabled. Enable it with --enable-sdk-logging on the command line or logging::enable-sdk-logging in your configuration file
2022-03-25T13:32:31.113Z [DEBUG] {Retry.cpp}: Retryable function starting, it will retry until success
2022-03-25T13:32:31.116Z [INFO]  {SharedCrtResourceManager.cpp}: Establishing MQTT connection with client id AWS_000251...
2022-03-25T13:32:31.267Z [INFO]  {SharedCrtResourceManager.cpp}: MQTT connection established with return code: 0
2022-03-25T13:32:31.267Z [INFO]  {SharedCrtResourceManager.cpp}: Shared MQTT connection is ready!
2022-03-25T13:32:31.267Z [INFO]  {Main.cpp}: Jobs is disabled
2022-03-25T13:32:31.267Z [INFO]  {Main.cpp}: Secure Tunneling is enabled
2022-03-25T13:32:31.267Z [INFO]  {Main.cpp}: Device Defender is disabled
2022-03-25T13:32:31.267Z [INFO]  {Main.cpp}: Sample shadow is disabled
2022-03-25T13:32:31.267Z [INFO]  {Main.cpp}: Pub Sub is disabled
2022-03-25T13:32:31.267Z [INFO]  {SharedCrtResourceManager.cpp}: Starting Device Client features.
2022-03-25T13:32:31.267Z [INFO]  {SecureTunnelingFeature.cpp}: Running Secure Tunneling!
2022-03-25T13:32:31.267Z [INFO]  {Main.cpp}: Client base has been notified that Secure Tunneling has started
2022-03-25T13:32:31.311Z [DEBUG] {SecureTunnelingFeature.cpp}: Subscribed to tunnel notification topic`

**Comment by shangabl** [comment]
Hi Anthony,

Sorry for the delay! I will take a look now and see.

**Comment by shangabl** [comment]
Hi Anthony,

Is it the source connection state or the destination connection state that is DISCONNECTED?

If it is the source then that is expected. You need to connect to the tunnel from the source device in order using the source access token.

**Comment by shangabl** [comment]
Hi Anthony, I'm going to close this out for now but feel free to reach out if you need any more help.

**Comment by ramg44** [comment]
Is everyone using the certs generated  while creating the iot thing right?
